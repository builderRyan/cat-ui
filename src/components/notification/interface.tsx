export type Types = 'default' | 'success' | 'warn' | 'error';
