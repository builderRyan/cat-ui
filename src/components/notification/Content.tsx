import React from 'react';
import classNames from 'classnames';
import Icon from '../Icon';
import PurePortal from '../PurePortal';
import PropTypes from 'prop-types';
import { CSSTransition } from 'react-transition-group';

import { Types } from './interface';

const TIMEOUT = 300; // ms

export interface ContentProps {
  id?: string;
  visible?: boolean;
  prefix?: string;
  className?: string;
  style?: React.CSSProperties;
  title?: React.ReactNode;
  content?: React.ReactNode;
  children?: React.ReactNode;
  type?: Types;
  selector?: string;
  onClose?: (id?: string) => void;
  iconStyle?: React.CSSProperties;
}

export interface ContentState {
  visible: boolean;
}

class Content extends React.Component<ContentProps, ContentState> {
  static displayName = 'Notification';

  static propTypes = {
    id: PropTypes.string,
    visible: PropTypes.bool,
    prefix: PropTypes.string,
    className: PropTypes.string,
    style: PropTypes.object,
    title: PropTypes.node,
    content: PropTypes.node,
    type: PropTypes.oneOf(['default', 'success', 'warn', 'error']),
    selector: PropTypes.string,
    onClose: PropTypes.func,
  }

  static defaultProps = {
    prefix: 'cat',
    selector: '.cat-notification-container',
    type: 'default'
  };

  static getDerivedStateFromProps(props: ContentProps) {
    if ('visible' in props) {
      return {
        visible: props.visible
      }
    }
    return null;
  }

  state = {
    visible: true,
  };

  handleClose = () => {
    this.setState({ visible: false });
    const { onClose, id } = this.props;
    onClose && onClose(id);
  };

  render() {
    const {
      prefix,
      title,
      content,
      className,
      type,
      style,
      selector,
      children,
      iconStyle
    } = this.props;
    const classPrefix = `${prefix}-notification`;
    const { visible } = this.state;

    const notificationContent = children || content;

    return (
      <PurePortal selector={selector}>
        <CSSTransition
          timeout={TIMEOUT}
          in={visible}
          classNames={classPrefix}
          unmountOnExit
          mountOnEnter
          appear
        >
          <>
            <div
              className={classNames(
                classPrefix,
                `${classPrefix}--${type}`,
                className
              )}
              style={style}
            >
              {title && <p className={`${classPrefix}__title`}>{title}</p>}
              {notificationContent && <div className={`${classPrefix}__content`}>{notificationContent}</div>}
              <Icon
                type='close'
                className={`${classPrefix}__close`}
                onClick={this.handleClose}
                style={iconStyle}
              />
            </div>
            <br />
          </>
        </CSSTransition>
      </PurePortal>
    );
  }
}

export default Content;
