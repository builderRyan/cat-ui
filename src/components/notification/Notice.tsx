import * as React from 'react';
import * as ReactDOM from 'react-dom';

interface NoticeOptions {
  /**
   * 容器类名
   */
  className: string;
  /**
   * 动画时间
   */
  animateTime: number;
}

class Notice {
  constructor(options: NoticeOptions) {
    this.options = options;
    const { className } = options;
    const container = document.querySelector(className);
    if (!container) {
      const div = document.createElement('div');
      div.className = className;
      document.body.appendChild(div);
    }
  }

  options: NoticeOptions;

  container: HTMLDivElement;
  index = 0;

  containerMaps: object = {};

  createContainerId = () => {
    return ++this.index;
  };

  remove = (id: string) => {
    const containerObj = this.containerMaps[id];
    if (!containerObj) {
      return;
    }
    const { container } = containerObj;
    setTimeout(() => {
      ReactDOM.unmountComponentAtNode(container);
      delete this.containerMaps[id];
    }, this.options.animateTime * 1000);
  };

  render(Component: any, props: any) {
    const div = document.createElement('div');

    const id = this.createContainerId();

    const cProps = {
      ...props,
      id
    };

    this.containerMaps[id] = { container: div };

    ReactDOM.render(<Component {...cProps} />, div);
  }
}

export default Notice;
