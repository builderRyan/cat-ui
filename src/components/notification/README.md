# notification 通知

## Example

```jsx
notification.success({
  title: 'title',
  content: 'content'
});
```

## API

- notification.default(config)
- notification.success(config)
- notification.warn(config)
- notification.error(config)

config 参数如下：

|  属性   |       类型        | 默认值 | 描述 | 是否必填 |
|:-------:|:-----------------:|:------:|:----:|:--------:|
|  title  | `React.ReactNode` |   -    | 标题 |    -     |
| content | `React.ReactNode` |   -    | 内容 |    -     |
