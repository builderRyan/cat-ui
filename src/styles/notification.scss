.cat-notification-container {
  position: fixed;
  z-index: 9999;
  margin: 0 20px 12px 0;
  right: 20px;
  bottom: 0;
}

.cat-notification {
  padding: 24px 66px 24px 24px;
  border-radius: 8px;
  color: #fff;
  z-index: 9999;
  position: relative;
  margin-bottom: 12px;
  transition: 0.3s;

  &-appear {
    transform: translateX(100%);
    &-active {
      transform: translateX(0);
    }
    &-done {
      transform: translateX(0);
    }
  }
  &-exit {
    transform: translateX(0);
    &-active {
      transform: translateX(100%);
    }
    &-done {
      transform: translateX(100%);
    }
  }

  &__title {
    font-size: 16px;
    font-weight: bold;
  }
  &__content {
    font-size: 14px;
    max-width: 254px;
  }
  &__close {
    position: absolute;
    right: 24px;
    top: 24px;
    cursor: pointer;
  }
  &--success {
    box-shadow: 0 4px 12px 0 rgba(0, 153, 255, 0.2);
    background-color: rgba(0, 153, 255, 0.9);
  }
  &--warn {
    box-shadow: 0 4px 12px 0 rgba(255, 102, 0, 0.2);
    background-color: rgba(255, 102, 0, 0.9);
  }
  &--error {
    box-shadow: 0 2px 8px 0 rgba(255, 108, 108, 0.3);
    background-color: rgba(255, 114, 110, 0.9);
  }
}
