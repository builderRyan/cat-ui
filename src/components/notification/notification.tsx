import NotificationContainer from '@utils/NotificationContainer';
import Content from './Content';

export type Types = 'success' | 'warn' | 'error';

export interface NotificationConfig {
  title?: React.ReactNode;
  content?: React.ReactNode;
}

export interface NotificationInterface {
  open: (type: Types, config: NotificationConfig) => void;
  success: (config: NotificationConfig) => void;
  warn: (config: NotificationConfig) => void;
  error: (config: NotificationConfig) => void;
}

let notificationInstance: any;
function getNotificationInstance() {
  if (notificationInstance) {
    return notificationInstance;
  }
  notificationInstance = new NotificationContainer({
    className: 'cat-notification-container',
    animateTime: 0.3
  });
  return notificationInstance;
}

function open(type: Types, config: NotificationConfig) {
  const notificationInstance = getNotificationInstance();
  const props = {
    type,
    title: config.title,
    content: config.content,
    onClose: notificationInstance.remove,
    selector: '.cat-notification-container'
  };
  notificationInstance.render(Content, props);
}

const notification: NotificationInterface = {
  open,
  success(config: NotificationConfig) {
    this.open('success', config);
  },
  warn(config: NotificationConfig) {
    this.open('warn', config);
  },
  error(config: NotificationConfig) {
    this.open('error', config);
  }
};

export default notification;
