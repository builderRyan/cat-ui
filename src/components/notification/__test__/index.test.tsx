import React from 'react';
import notification from '../index';
import { mount } from 'enzyme';
import sinon from 'sinon';


const { Notification } = notification;

describe('notification', () => {
  test('renders one notification', () => {
    notification.success({
      title: 'success title',
      content: 'success content'
    });
    const dom: any = document.querySelector('.cat-notification-container');
    expect(dom.childNodes.length).toBe(2);
  });

  test('renders two notification', () => {
    notification.warn({
      title: 'warn title',
      content: 'warn content'
    });
    const dom: any = document.querySelector('.cat-notification-container');
    expect(dom.childNodes.length).toBe(4);
  });

  test('renders three notification', () => {
    notification.error({
      title: 'error title',
      content: 'error content'
    });
    const dom: any = document.querySelector('.cat-notification-container');
    expect(dom.childNodes.length).toBe(6);
  });

  test('renders four notification', () => {
    notification.default({
      title: 'defaut title',
      content: 'default content'
    });
    const dom: any = document.querySelector('.cat-notification-container');
    expect(dom.childNodes.length).toBe(8);
  });
});


describe('Notification', () => {
  test('render Notification', () => {
    document.body.innerHTML = '';
    const handleClose = sinon.spy();

    const wrapper = mount(
      <Notification visible={false} title='default title' onClose={handleClose}>
        default content
      </Notification>
    );

    expect(wrapper).toMatchSnapshot();
  });
});
