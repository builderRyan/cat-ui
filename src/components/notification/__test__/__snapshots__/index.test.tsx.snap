// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification render Notification 1`] = `
<withCreateNotificationContainer(Notification)
  onClose={[Function]}
  title="default title"
  visible={false}
>
  <Notification
    onClose={[Function]}
    prefix="cat"
    selector=".cat-notification-container"
    title="default title"
    type="default"
    visible={false}
  >
    <PurePortal
      selector=".cat-notification-container"
    >
      <Portal
        containerInfo={
          <div
            class="cat-notification-container"
          />
        }
      >
        <CSSTransition
          appear={true}
          classNames="cat-notification"
          in={false}
          mountOnEnter={true}
          timeout={300}
          unmountOnExit={true}
        >
          <Transition
            appear={true}
            enter={true}
            exit={true}
            in={false}
            mountOnEnter={true}
            onEnter={[Function]}
            onEntered={[Function]}
            onEntering={[Function]}
            onExit={[Function]}
            onExited={[Function]}
            onExiting={[Function]}
            timeout={300}
            unmountOnExit={true}
          />
        </CSSTransition>
      </Portal>
    </PurePortal>
  </Notification>
</withCreateNotificationContainer(Notification)>
`;
